[SPARK-42534][SQL] Fix DB2Dialect Limit clause

sadikovi opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

The PR fixes DB2 Limit clause syntax. Although DB2 supports LIMIT keyword, it seems that this support varies across databases and versions and the recommended way is to use `FETCH FIRST x ROWS ONLY`. In fact, some versions don't support LIMIT at all. Doc: https://www.ibm.com/docs/en/db2/11.5?topic=subselect-fetch-clause, usage example: https://www.mullinsconsulting.com/dbu_0502.htm.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

Fixes the incorrect Limit clause which could cause errors when using against DB2 versions that don't support LIMIT.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

No.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->

I added a unit test and an integration test to cover this functionality.

Commit:
update

Commit:
update

sadikovi commented:
cc @dongjoon-hyun @cloud-fan

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:
```diff
@@ -1028,6 +1028,19 @@ class JDBCSuite extends QueryTest with SharedSparkSession {
+        .withLimit(123)
+        .build()
+        .trim() ==
+      "SELECT a,b FROM test     FETCH FIRST 123 ROWS ONLY")
```
very nit: do you know why the generated SQL has more than one spaces before the limit clause?

sadikovi replied:
Yes, I do. This is because of the string interpolation for the JDBC query that is written like `"$a $b $c...`. So when some of the components are empty, whitespaces are still kept between the components resulting in the long gaps.

```scala
options.prepareQuery +
      s"SELECT $myTopExpression $columnList FROM ${options.tableOrQuery} $myTableSampleClause" +
      s" $myWhereClause $getGroupByClause $getOrderByClause $myLimitClause $myOffsetClause"
```

https://github.com/apache/spark/blob/4b50a46f1c6ba4ffe2c42f70a512879c28d11dcf/sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcSQLQueryBuilder.scala#L163.

dongjoon-hyun commented on connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DB2IntegrationSuite.scala:
```diff
@@ -217,4 +217,26 @@ class DB2IntegrationSuite extends DockerJDBCIntegrationSuite {
     assert(actual.toSet === expectedResult)
   }
+
+  test("SPARK-42534") {
```
Could you add some description, please? :)

sadikovi replied:
Absolutely 👍.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/DB2Dialect.scala:
```diff
@@ -160,4 +160,8 @@ private object DB2Dialect extends JdbcDialect {
   }
+
+  override def getLimitClause(limit: Integer): String = {
+    if (limit > 0 ) s"FETCH FIRST $limit ROWS ONLY" else ""
```
nit. Space.
```scala
- if (limit > 0 )
+ if (limit > 0)
```

dongjoon-hyun commented on connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DB2IntegrationSuite.scala:
```diff
@@ -217,4 +217,26 @@ class DB2IntegrationSuite extends DockerJDBCIntegrationSuite {
+      .load()
+      .collect()
+
+    assert(actual.length === 2)
```
nit. The next `assert` looks inclusive and enough. Shall we remove this?

sadikovi replied:
Good point, I will remove.

dongjoon-hyun reviewed (commented):
I have three minor comments, @sadikovi .

Commit:
address comments

sadikovi commented:
@dongjoon-hyun I have addressed the comment, could you review again please? Thank you.

Also, do you know whom I can ping on this PR with the regard to DB2 SQL semantics?

dongjoon-hyun reviewed (approved):
+1 from my side. Thank you for update, @sadikovi .

For DB2 questions, I believe @huaxingao and @dilipbiswal have some experience from their previous company, IBM.

huaxingao commented on connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DB2IntegrationSuite.scala:
```diff
@@ -217,4 +217,25 @@ class DB2IntegrationSuite extends DockerJDBCIntegrationSuite {
+      .option("url", jdbcUrl)
+      .option("dbtable", "tbl")
+      .load()
+      .limit(2)
```
Limit push down is only supported in V2 JDBC. I don't think this `limit(2)` will trigger the newly added `DB2Dialect.getLimitClause`. This test should be done in `v2/DB2IntegrationSuite`

sadikovi replied:
There are already tests covering limit pushdown.

cloud-fan replied:
I think it's OK to leave the test here, in case one day we support limit pushdown in v1 JDBC as well.

sadikovi commented:
@dongjoon-hyun Is there anything else left on this PR to merge?

cloud-fan commented:
thanks, merging to master/3.4 (bug fix)!

cloud-fan commented:
it has conflicts with 3.4, @sadikovi can you open a backport PR? thanks!

sadikovi commented:
Yes, I will open a separate PR.

sadikovi commented:
PR for 3.4: https://github.com/apache/spark/pull/40155.
